TreasureData.agent.lib.nut

```diff
@@ -0,0 +1,165 @@
+            _sendHeaders,
+            data,
+            function (error) {
+                _invokeDefaultCallback(callback, error, data);
+            }.bindenv(this));
+    }
+
+    // Enables/disables the library debug output (including errors logging).
+    // Disabled by default.
+    //
+    // Parameters:
+    //     value : boolean             true to enable, false to disable
+    function setDebug(value) {
+        _debug = value;
+    }
+
+    // -------------------- PRIVATE METHODS -------------------- //
+
+    // Sends an http request to TreasureData.
+    function _processRequest(method, endpoint, path, headers, body, callback) {
+        local url = endpoint + path;
+        local encodedBody = http.jsonencode(body);
+        _logDebug(format("Doing the request: %s %s, body: %s", method, url, encodedBody));
+
+        local request = http.request(method, url, headers, encodedBody);
+        request.sendasync(function (response) {
+            _processResponse(response, callback);
+        }.bindenv(this));
+    }
+
+    // Processes http response from TreasureData and executes callback.
+    function _processResponse(response, callback) {
+        _logDebug(format("Response status: %d, body: %s", response.statuscode, response.body));
+        local error = null;
+        local httpStatus = response.statuscode;
+        if (httpStatus < 200 || httpStatus >= 300) {
+            local body = null;
+            try {
+                body = (response.body == "") ? {} : http.jsondecode(response.body);
+            } catch (e) {
+                _logError(e);
+            }
+            error = TreasureData.Error(httpStatus, body);
+        }
+        callback(error);
+    }
+
+    // Logs error occurred and invokes callback with default parameters.
+    function _invokeDefaultCallback(callback, error, data) {
+        if (error) {
+            _logError(format("%s: %i", _TREASURE_DATA_REQUEST_FAILED, error.httpStatus));
+        }
+        if (callback) {
+            imp.wakeup(0, function () {
+                callback(error, data);
+            });
+        }
+    }
+
+    // Logs error occurred during the library methods execution.
+    function _logError(message) {
+        if (_debug) {
+            server.error("[TreasureData] " + message);
+        }
+    }
+
+    // Logs debug messages occurred during the library methods execution.
+    function _logDebug(message) {
+        if (_debug) {
+            server.log("[TreasureData] " + message);
+        }
+    }
+
+    // Auxiliary class, represents an error returned by the library when an HTTP request to the Treasure Data platform fails.
+    // The error details can be found in the error.httpStatus and error.httpResponse properties.
+    Error = class {
+        // HTTP status code (integer)
+        httpStatus = null;
+
+        // Response body of the failed request (table)
+        httpResponse = null;
+
+        constructor(httpStatus, httpResponse) {
```

Comment:
Why don't we just use the native `httpresponse` object instead? it has statuscode, body, etc. I would just probably eliminate this `Error` class unless there is a strong reason to have it.